lib/src/analysis_options.dart

```diff
@@ -3,79 +3,102 @@
-
-String _analysisOptions(String pedanticConfigPath) => '''
-# Defines a default set of lint rules enforced for
-# projects at Google. For details and rationale,
-# see https://github.com/dart-lang/pedantic#enabled-lints.
-include: $pedanticConfigPath
-
-# For lint rules and documentation, see http://dart-lang.github.io/linter/lints.
-# Uncomment to specify additional rules.
-# linter:
-#   rules:
-#     - camel_case_types
-
-# analyzer:
-#   exclude:
-#     - path/to/excluded/files/**
-''';
-
-// Keep it updated with
-// https://github.com/flutter/flutter/blob/master/packages/flutter/lib/analysis_options_user.yaml
-const String _flutterAnalysisOptions = '''
-analyzer:
-  errors:
-    # treat missing required parameters as a warning (not a hint)
-    missing_required_param: warning
-
-# Source of linter options:
-# https://dart-lang.github.io/linter/lints/options/options.html
-
-linter:
-  rules:
-    - avoid_empty_else
-    - avoid_init_to_null
-    - avoid_return_types_on_setters
-    - await_only_futures
-    - camel_case_types
-    - cancel_subscriptions
-    - close_sinks
-    - control_flow_in_finally
-    - empty_constructor_bodies
-    - empty_statements
-    - hash_and_equals
-    - implementation_imports
-    - library_names
-    - non_constant_identifier_names
-    - package_api_docs
-    - package_names
-    - package_prefixed_library_names
-    - prefer_is_not_empty
-    - slash_for_doc_comments
-    - super_goes_last
-    - test_types_in_equals
-    - throw_in_finally
-    - type_init_formals
-    - unnecessary_brace_in_string_interps
-    - unnecessary_getters_setters
-    - unnecessary_statements
-    - unrelated_type_equality_checks
-    - valid_regexps
-''';
+final _logger = Logger('analysis_options');
+
+String _cachedFlutterOptionsOnGithub;
+String _cachedPedanticOptionsOnGithub;
+
+/// Returns the default analysis options (in yaml format).
+Future<String> getDefaultAnalysisOptionsYaml({
+  @required bool usesFlutter,
+  @required String flutterSdkDir,
+}) async {
+  if (usesFlutter) {
+    return await _getFlutterAnalysisOptions(flutterSdkDir);
+  } else {
+    return await _getPedanticAnalysisOptions();
+  }
+}
+
+Future<String> _getFlutterAnalysisOptions(String flutterSdkDir) async {
+  // try to load local file
+  flutterSdkDir ??= Platform.environment['FLUTTER_ROOT'];
+  if (flutterSdkDir != null &&
+      flutterSdkDir.isNotEmpty &&
+      await Directory(flutterSdkDir).exists()) {
+    final file = File(p.join(flutterSdkDir, 'packages', 'flutter', 'lib',
+        'analysis_options_user.yaml'));
+    if (await file.exists()) {
+      return await file.readAsString();
+    }
+  }
+
+  // try to load latest from github
+  if (_cachedFlutterOptionsOnGithub != null) {
+    return _cachedFlutterOptionsOnGithub;
+  }
+  try {
+    final rs = await http.get(Uri.parse(
+        'https://raw.githubusercontent.com/flutter/flutter/master/packages/flutter/lib/analysis_options_user.yaml'));
```

Comment:
When and why do we need this? It's only relevant when the flutter SDK isn't available, right?
Why do we need to support that use case?

-------

Also HTTP requests without retries is never a good idea.

Reply:
> When and why do we need this?

pub.dev doesn't need this, only standalone CI run without explicit Flutter SDK directory set. (We could detect where it is in the path and work out the subdirectory, but it may have issues on Windows).

Reply:
Okay, I'm down for that...

Let's add some retries through... or use `http_retry`.

Reply:
Implemented with hand-rolled retry, because `http_retry` will conflict with `pub-dev` (at least for a while).